Fixed #31147 -- Added SearchHeadline to django.contrib.postgres.

hannseman opened the pull request:
Ticket: https://code.djangoproject.com/ticket/31147

felixxm commented on django/contrib/postgres/search.py:
```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
+            query = SearchQuery(query)
+        self.config = query.config
+        self.options = extra.get('options', self.options)
+        super().__init__(expression, query, **extra)
```
Can we validate at least names of options?

hannseman replied:
Sure, in my first draft I actually remapped all the CamelCased postgres option names to more pythonic names:

```python
option_mapping = {
    'start_sel': 'StartSel',
    'stop_sel': 'StopSel',
    'max_words': 'MaxWords',
    'min_words': 'MinWords',
    'short_word': 'ShortWord',
    'highlight_all': 'HighlightAll',
    'max_fragments': 'MaxFragments',
    'fragment_delimiter': 'FragmentDelimiter',
}
```

But then it felt like the need to more of less duplicate the documentation from the postgres docs for our version of the options would end up being a little bit too much. 

Do you think we should use the postgres names and validate them or keep our own mapping?

felixxm replied:
IMO we should check options with PostreSQL names.

hannseman commented on django/contrib/postgres/search.py:
```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
+class SearchHeadline(Func):
+    function = 'ts_headline'
+    output_field = TextField()
+    options = None
```
My intention was to make it easy to set your own defaults for `options` by extending `SearchHeadline`:

```python
class CustomSearchHeadline(SearchHeadline):
    options = {
        'StartSel': '<strong>',
        'StopSel': '</strong>'
    }
```

Much like how `SearchVector.config` works: https://github.com/django/django/blob/f14b13b832fd5ecedc6bac94da3e77fc188193a8/django/contrib/postgres/search.py#L57

But I just realised that if we would choose to keep the `options` class attribute for this purpose we should not override them with the `options` passed as an argument. Instead we could do something like this:
```python
class SearchHeadline(Func):
    ...
    options = {}
    def __init__(self, expression, query, **extra):
        ...
        options = extra.get('options') or {}
        self.options = {**self.options, **options}
```

hannseman commented on django/contrib/postgres/search.py:
```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
+    def __init__(self, expression, query, **extra):
+        if not hasattr(query, 'resolve_expression'):
+            query = SearchQuery(query)
+        self.config = query.config
```
I couldn't really see the point of having to repeat the config when it's already set on the `SearchQuery`. And if I'm not missing something here specifying two different configs could give pretty weird results:

```python
SearchHeadline('dialogue', SearchQuery('cadeaux', config='french'), config='chinese')
```

felixxm replied:
So do we need to generate SQL with the same config in the `ts_headline` and an inline `tsquery`? (that's my main question)
```sql
ts_headline('french'::regconfig,  ..., tsquery(..., 'french'::regconfig), ...)
```
If not then I would prefer to leave config only in `tsquery` and don't duplicate its logic in `SearchHeadline`.

hannseman replied:
Yes AFAIK the config needs to be explicitly specified on `ts_headline` even if it already is set on the `tsquery`. If not specified it will fall back on the config set as `default_text_search_config`.

felixxm replied:
I think we should support both, the main `config` is used for parsing a document and `tsquery`'s `config` is used for parsing a query. I would handle the main `config` in `SearchHeadline` and leave `SearchQuery()` as a expression.

hannseman replied:
Sure that makes sense. Maybe we should always allow to set `config` which will override config from an eventual `SearchQuery ` when set?

hannseman commented on django/contrib/postgres/search.py:
```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
+            if self.options:
+                options = []
+                for option, value in self.options.items():
+                    pg_value = psycopg2.extensions.adapt(value).getquoted().decode()
```
I might be misunderstanding something but the problem here is that the whole `options` string needs to be passed as a value to `execute`. Otherwise psycopg2 will apply invalid quoting, for example:

```python
 cursor.execute(".., '%s=%s, %s=%s'", ('StartSel', '<strong>', 'StopSel', '</strong>'))
```
*Note the required quoting around the options template string*

Would quote `StartSel` and `StopSel` which is invalid syntax: 
```
''StartSel'='<strong>', 'StopSel'='</strong>''
```

Using `psycopg2.extensions.adapt` allows us to apply the correct quoting for strings and proper conversion of booleans.

If we where to still wrap value with `Value` it would have to be used something like this:
```python
_, value_params = compiler.compile(value)
pg_value = psycopg2.extensions.adapt(value_params[0]).getquoted().decode()
 ```

And that feels kind of awkward.

Let me know if I'm missing something!

felixxm replied:
@charettes Ideas welcome. Can you take a look? I really don't like the idea of using `psycopg2` in expressions.

charettes replied:
@felixxm, @hannseman is right we need to compile a single `text` value for all the options so I don't see a way around invoking `adapt` here like we do when _baking_ a `DEFAULT` string value

https://github.com/django/django/blob/22ce5d0031bd795ade081394043833e82046016c/django/db/backends/postgresql/schema.py#L34-L38

hannseman commented on django/contrib/postgres/search.py:
```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
+                resolved.config = Value(self.config).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+            else:
+                resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        return resolved
```
Good idea!

felixxm replied:
I wonder if we could add a `SearchConfig` expression :thinking: (maybe a subclass of `Value`):
```python
class SearchConfig(Expression):
    def __init__(self, config):
        super().__init__(output_field=None)
        ...

    def as_sql(self, compiler, connection):
        ...
        return '%::regconfig', params
```
This should simplify everything.

hannseman replied:
I guess something like this could work:

```python
class SearchConfig(Expression):
    def __init__(self, config):
        super().__init__(output_field=None)
        self.config = config

    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
        if self.config:
            if not hasattr(self.config, 'resolve_expression'):
                resolved.config = Value(self.config).resolve_expression(query, allow_joins, reuse, summarize, for_save)
            else:
                resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
        return resolved

    def as_sql(self, compiler, connection):
        sql, params = compiler.compile(self.config)
        sql = '{}::regconfig'.format(sql)
        return sql, params
```

Not really sure what the point of basing this class of `Value` would be.

Note that every class using `SearchConfig` would need to implement `resolve_expression` like this (please correct me if I'm missing something here):

```python
class SearchHeadline(Func):
    def __init__(self, expression, query, config=None, **extra):
        ...
        self.config = SearchConfig(config) if config else None

    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
        if self.config:
            resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
        return resolved
    
    ...
```

It could then be used in `as_sql` as something like this:
```python
def as_sql(self, compiler, connection, ...):
    ...
    if self.config:
        config_sql, config_params = compiler.compile(self.config)
        extra_context['config'] = '{}, '.format(config_sql)
```

As it still would require overriding `resolve_expression` (we still could use a mixin but simpler) it would still be a bit cumbersome but I guess that not having to handle the casting to `::regconfig` everywhere weight heavier. What do you think?

hannseman replied:
I pushed a version which introduces `SearchConfig` as an expression (https://github.com/django/django/pull/12297/commits/709e36b5df769a169df7953cf0b54c81e4743b08). I prefer it over the mixin approach. Let me know what you think. If you agree I'll squash it into https://github.com/django/django/pull/12297/commits/96b55ce9c9c72d4f5fcde4f9eb85425426201266 and if not we can just revert it.

felixxm replied:
@hannseman Thanks :+1: 

>  I prefer it over the mixin approach.

Yes me too :+1: . We can move `Value()` wrapping to the `__init__()` and simplify it a bit, e.g.:
```python
class SearchConfig(Expression):
    def __init__(self, config):
        super().__init__(output_field=None)
        if not (config and hasattr(config, 'resolve_expression')):
            config = Value(config)
        self.config = config
                          
    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
        resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
        return resolved
        
    def as_sql(self, compiler, connection):
        sql, params = compiler.compile(self.config)
        return '%s::regconfig' % sql, params
```

Please move introducing a `SearchConfig` expression to the separate commit, or even PR.

hannseman replied:
Much cleaner with the `Value` in `__init__()` 👍 

I opened up a new PR for this: https://github.com/django/django/pull/12377

I'll rebase this PR once https://github.com/django/django/pull/12377 is merged.

felixxm replied:
Merged :confetti_ball:

hannseman replied:
Rebased :raised_hands:

charettes commented on django/contrib/postgres/search.py:
```diff
@@ -240,6 +242,65 @@ def as_sql(self, compiler, connection, function=None, template=None):
+        'HighlightAll',
+        'MaxFragments',
+        'FragmentDelimiter',
+    )
```
Should be a `frozenset` since it's only used for containment. From that point you use `set(options).difference(self.VALID_OPTIONS)` to determine `invalid_options`.

charettes commented on tests/postgres_tests/test_search.py:
```diff
@@ -453,3 +459,111 @@ def test_str(self):
+                options={
+                    'InvalidOption': 'foo',
+                    'AnotherInvalidOption': 'bar',
+                    'MaxFragments': 4
```
This test is really critical because without the whitelisting injection could take via malformed keys.

hannseman replied:
Yes it's a bit scary. Let me know if you got any ideas to make it more solid. Maybe adding a comment to not have it accidentally removed could help?

charettes commented on django/contrib/postgres/search.py:
```diff
@@ -240,6 +242,65 @@ def as_sql(self, compiler, connection, function=None, template=None):
+                for option, value in self.options.items():
+                    pg_value = psycopg2.extensions.adapt(value).getquoted().decode()
+                    options.append('%s=%s' % (option, pg_value))
+                options_params = [', '.join(options)]
```
append instead of creating a new list

```suggestion
                options_params.append(', '.join(options))
```

All off the above could also be reduced to

```python
 options_params.append(', '.join(
    '%s=%s' % (option, psycopg2.extensions.adapt(value).getquoted().decode())
    for option, value in options.items()
))

hannseman replied:
Much nicer, thanks!

charettes commented on django/contrib/postgres/search.py:
```diff
@@ -240,6 +242,65 @@ def as_sql(self, compiler, connection, function=None, template=None):
+        config_params, options_params = [], []
+        extra_context = {}
+        if template is None:
+            template = '%(function)s(%(config)s%(expressions)s%(options)s)'
```
What's the reason for branching off `template is None` here instead of defining this template as a class attribute and moving along like other normal `Func`?

hannseman replied:
I guess I looked at the other classes in `django.contrib.postgres.search` and saw the pattern checking `template is None`  but misunderstood its usage. Thanks!

charettes commented on django/contrib/postgres/search.py:
```diff
@@ -240,6 +242,65 @@ def as_sql(self, compiler, connection, function=None, template=None):
+
+    def as_sql(self, compiler, connection, function=None, template=None, arg_joiner=None, **extra_context):
+        config_params, options_params = [], []
+        extra_context = {}
```
Why are you completely overriding the provided kwarg provided `extra_context`?

hannseman replied:
A complete mistake. Thanks!

hannseman commented:
> I personally think that accepting options as __init__(kwargs) (e.g. start_sel) would make for a nicer interface though.

@charettes We touched on providing nicer snake_cased names for the options https://github.com/django/django/pull/12297#discussion_r365405971. Although this was for keeping them in the `options`-kwarg and not as separate kwargs. 

I agree that it would be a nicer interface but apart from the issue of requiring documentation for all options one would not have the same convenience of passing a shared configuration like:
```python
SearchHeadline(
    'dialogue',
    SearchQuery('brave sir robin'),
    options=settings.HEADLINE_CONFIG,
)
```

But if we decide to keep the class-attribute of `SearchHeadline.options` as discussed in https://github.com/django/django/pull/12297#discussion_r365402712 one could set a shared configuration by `extending` SearchHeadline and overriding `options`.

charettes commented:
> I agree that it would be a nicer interface but apart from the issue of requiring documentation for all options one would not have the same convenience of passing a shared configuration like:

I agree about the documentation needs but the shared configuration point is moot as you could pass `**settings.HEADLINE_CONFIG` given your use the documented kwargs keys.

charettes commented:
> But if we decide to keep the class-attribute of SearchHeadline.options as discussed in #12297 (comment) one could set a shared configuration by extending SearchHeadline and overriding options.

One could do as well by overriding `__init__` as long as options are stored at the instance level.

hannseman commented:
> I agree about the documentation needs but the shared configuration point is moot as you could pass **settings.HEADLINE_CONFIG given your use the documented kwargs keys.

True.

> One could do as well by overriding __init__ as long as options are stored at the instance level. 

You're right, I looked at how `django.contrib.postgres.search.SearchVector` kept `config` as a class-attribute but passing configuration as `options=settings.HEADLINE_CONFIG` or `**settings.HEADLINE_CONFIG` should be enough. 

I've removed the `options` class-attribute.

hannseman commented:
@felixxm do you have any thoughts about accepting options as separate kwargs instead of passing a dictionary to the `options` argument?

felixxm commented:
@hannseman I would prefer a separate kwargs, this is a safer option (e.g. `.., *, start_sel=None, stop_sel=None, max_words=None, min_words=None, short_word=None, highlight_all=None, max_fragments=None, fragment_delimiter=None)`).

hannseman commented:
@felixxm Thanks for the quick reply! 

I've rebased with a new implementation of options as separate kwargs and added documentation for them. Using kwargs feels a lot more solid 👍

felixxm commented:
@hannseman Thanks for all updates :+1: I reverted unrelated changes, pushed edits to docs, used more specific assertions in tests, and removed `SQL` assertions (it's enough to test results).

felixxm commented:
@charettes Thanks :+1: I pushed updates.

@carltongibson Thanks :+1:

hannseman commented:
@felixxm @charettes @carltongibson many thanks for the reviews and fixes 🎉
